### PR TITLE
Reauth reimplemented

### PR DESCRIPTION
**Pull request by alxdarksage**
Design: https://sagebionetworks.jira.com/wiki/spaces/BRIDGE/pages/755171447/Reauthentication+Improvements

JIRA: https://sagebionetworks.jira.com/browse/BRIDGE-2373

The following SQL will need to be run prior to deployment:

```SQL
CREATE TABLE `AccountSecrets` (
  `accountId` VARCHAR(255) NOT NULL,
  `hash` VARCHAR(255) NOT NULL,
  `type` ENUM('REAUTH') DEFAULT 'REAUTH',
  `algorithm` ENUM('STORMPATH_HMAC_SHA_256', 'BCRYPT', 'PBKDF2_HMAC_SHA_256') NOT NULL,
  `createdOn` BIGINT NOT NULL,
  PRIMARY KEY (`accountId`,`hash`),
  CONSTRAINT `fk_account_secrets` FOREIGN KEY (`accountId`) REFERENCES `Accounts` (`id`) ON DELETE CASCADE,  
  INDEX `secrets_idx` (`accountId`, `type`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
```

### CODE REVIEW

File: app/org/sagebionetworks/bridge/exceptions/ConcurrentModificationException.java [hunk 1]
```diff
@@ -7,7 +7,7 @@
 import org.sagebionetworks.bridge.models.BridgeEntity;
 
 @SuppressWarnings("serial")
-@NoStackTraceException
+//@NoStackTraceException
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/exceptions/ConcurrentModificationException.java]: I assume this is just for testing? We'll want to revert this (and in EntityNotFoundException). Otherwise, 409s and 404s get logged as errors.

File: app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java [hunk 2]
```diff
@@ -172,51 +179,59 @@ public Account reauthenticate(Study study, SignIn signIn) {
...
+        if (account.getReauthTokenHash() != null) {
+            // Rotate the token to the secrets table before verifying. This migrates the tables if
+            // a reauthentication call occurs before we run a migration on this record.
+            AccountSecret secret = AccountSecret.create();
+            secret.setAccountId(account.getId());
+            secret.setAlgorithm(account.getReauthTokenAlgorithm());
+            secret.setHash(account.getReauthTokenHash());
+            secret.setType(AccountSecretType.REAUTH);
+            secret.setCreatedOn(account.getReauthTokenModifiedOn());
+            hibernateHelper.create(secret, null);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java:192]: Should we be calling AccountSecretDao instead?

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java [hunk 3]
```diff
@@ -596,41 +567,75 @@ public void authenticateAccountHasPasswordAlgorithmNoHash() throws Exception {
...
-        verify(mockHibernateHelper).queryGet(expQuery, EMAIL_QUERY_PARAMS, null, null, HibernateAccount.class);
+        verify(mockHibernateHelper, times(2)).queryGet(expQuery, EMAIL_QUERY_PARAMS, null, null, HibernateAccount.class);
 
-        // We update the account with a reauthentication token
+        // We update the account twice in this scenario
         verify(mockHibernateHelper).update(hibernateAccount, null);
-        // The hash has been changed
-        assertNotEquals(originalReauthTokenHash, hibernateAccount.getReauthTokenHash());
-        // This has been hashed
-        assertNotEquals(account.getReauthToken(), hibernateAccount.getReauthTokenHash());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java:633]: Should we verify that verifySecret() is called? Should we verified that the Reauth token has been moved from the account to Secrets?

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java [hunk 4]
```diff
@@ -596,41 +567,75 @@ public void authenticateAccountHasPasswordAlgorithmNoHash() throws Exception {
...
+        
+        // verify query
+        verify(mockHibernateHelper).queryGet(expQuery, EMAIL_QUERY_PARAMS, null, null, HibernateAccount.class);
+
+        // We update the account twice in this scenario
+        verify(mockHibernateHelper, never()).update(any(), any());
+        // These fields have remained null.
+        assertNull(hibernateAccount.getReauthTokenHash());
+        assertNull(hibernateAccount.getReauthTokenAlgorithm());
+        assertNull(hibernateAccount.getReauthTokenModifiedOn());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java:657]: Similarly here, we should also verify that verifySecret() is called.

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java [hunk 5]
```diff
@@ -682,15 +695,70 @@ public void reauthenticateAccountDisabled() throws Exception {
...
+        InOrder inOrder = Mockito.inOrder(mockHibernateHelper, mockAccountSecretDao);
+        
+        inOrder.verify(mockHibernateHelper).queryGet(any(), any(), any(), any(), any());
+        
+        inOrder.verify(mockHibernateHelper).create(secretCaptor.capture(), eq(null));
+        AccountSecret capturedSecret = secretCaptor.getValue();
+        assertEquals(hibernateAccount.getId(), capturedSecret.getAccountId());
+        assertNotNull(capturedSecret.getAlgorithm());
+        assertNotNull(capturedSecret.getCreatedOn());
+        assertNotNull(capturedSecret.getHash());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Instead of merely checking not null, we should check that it's the same as the ones we wrote into the account.

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java [hunk 6]
```diff
@@ -682,15 +695,70 @@ public void reauthenticateAccountDisabled() throws Exception {
...
+                ImmutableList.of(makeValidHibernateAccount(false)));
+        
+        // It also has no record in the secrets table...
+        
         // execute
         dao.reauthenticate(study, REAUTH_SIGNIN);
     }
     
+    @Test
+    public void reauthenticateMigratesReauthTokenFromAccount() throws Exception {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Is this test redundant with reauthenticateSuccessWithMigration()?

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java [hunk 7]
```diff
@@ -682,15 +695,70 @@ public void reauthenticateAccountDisabled() throws Exception {
...
+        assertSame(hibernateAccount, accountCaptor.getValue());
+        assertNull(accountCaptor.getValue().getReauthTokenHash());
+        assertNull(accountCaptor.getValue().getReauthTokenAlgorithm());
+        assertNull(accountCaptor.getValue().getReauthTokenModifiedOn());
+        
+        inOrder.verify(mockAccountSecretDao).verifySecret(REAUTH, ACCOUNT_ID, REAUTH_TOKEN, 3);
+    }
+    
+    @Test
+    public void reauthenticateNoMigrationFromAccount() throws Exception {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Is this redundant with reauthenticateSuccessNoMigration()?

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java [hunk 8]
```diff
@@ -682,15 +695,70 @@ public void reauthenticateAccountDisabled() throws Exception {
...
+        when(mockHibernateHelper.queryGet(any(), any(), any(), any(), any()))
+                .thenReturn(ImmutableList.of(hibernateAccount));
+        
+        AccountSecret secret = AccountSecret.create();
+        when(mockAccountSecretDao.verifySecret(any(), any(), any(), any(Integer.class))).thenReturn(Optional.of(secret));
+        
+        dao.reauthenticate(study, REAUTH_SIGNIN);
+        
+        verify(mockHibernateHelper, never()).create(any(), any());
+        verify(mockHibernateHelper, never()).update(any(), any());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Should we verify verifySecret() is called?

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java [hunk 9]
```diff
@@ -167,10 +169,11 @@ public UserSession signIn(Study study, CriteriaContext context, SignIn signIn) t
         Validate.entityThrowingException(SignInValidator.PASSWORD_SIGNIN, signIn);
         
         Account account = accountDao.authenticate(study, signIn);
-
+        
+        clearSession(study.getStudyIdentifier(), account.getId());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:173]: Will this changelist include having a set of multiple valid sessions? If so, instead of invalidating sessions, should we wait until the user signs in before invalidating them?

File: test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java [hunk 10]
```diff
@@ -201,13 +205,11 @@ public void after() {
     
     @Test // Test some happy path stuff, like the correct initialization of the user session
     public void signIn() {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java:213]: Should this test also verify deleting the reauth token? (Similarly for emailSignIn and phoneSignIn?)

File: test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java [hunk 11]
```diff
@@ -432,10 +440,12 @@ public void emailSignInAdminOK() {
     
     @Test
     public void reauthentication() {
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java:453]: Should this test verify the logic for preserving session ID?

File: test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java [hunk 12]
```diff
@@ -1063,5 +1018,7 @@ public void getSessionFromAccount() {
         assertEquals(TestConstants.TEST_STUDY, session.getStudyIdentifier());
         assertEquals(REAUTH_TOKEN, session.getReauthToken());
         assertEquals(CONSENTED_STATUS_MAP, session.getConsentStatuses());
+        
+        verify(accountSecretDao).createSecret(AccountSecretType.REAUTH, USER_ID, REAUTH_TOKEN);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java:1054]: Should there be an equivalent test for when reauth enabled = false?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/exceptions/ConcurrentModificationException.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/exceptions/ConcurrentModificationException.java]: Uncommented this and EntityNotFoundException (sorry, it was for testing).

File: app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java (hunk 2 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java:192]: We can't because createSecret would re-hash the already hashed token. Because this is just for migration, I didn't want to add a special method in the DAO for this, but if you want I can add. I intend to migrate and then remove this code.

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java (hunk 3 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java:633]: Added a check to the migration path that the create method of the hibernate helper is called with a fully initialized secrets object, along with verifySecret(). In the no migration path, verify the create call isn't called and verifySecret() is called.

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java (hunk 4 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java:657]: See above (added).

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java (hunk 5 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Random salt was making this difficult... I just substituted a static value rather than actually hashing, now checking for that value.

Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: There are also two methods to test reauth with migration and two without... I have consolidated these.

Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Also changed all the modifiedOn tests to test for the exact time we're looking for, not just null.

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java (hunk 6 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Yes, consolidated these four tests into two tests.

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java (hunk 7 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Yes

File: test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java (hunk 8 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/hibernate/HibernateAccountDaoTest.java]: Added

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java (hunk 9 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/AuthenticationService.java:173]: It won't, but we could do this (another alternative would be a grace period where we return the same session token, but see discussion in response to your comment on the design doc. Let's discuss and file a JIRA what what we want to do, I can do it next).

File: test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java (hunk 10 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java:213]: Added, along with an inOrder mock to verify the clear happens before the set, to the signIn, emailSignIn and phoneSignIn main tests.

File: test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java (hunk 11 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java:453]: Added reauthenticationPersistsExistingSessionTokens() test of this specific scenario.

File: test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java (hunk 12 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/AuthenticationServiceMockTest.java:1054]: Added getSessionFromAccountWithoutReauthentication().

**Review by DwayneJengSage — APPROVED** [review]

**Review by DwayneJengSage — APPROVED** [review]
